src/ai/whylabs/templates/batch_bigquery_template.py

```diff
@@ -98,6 +134,82 @@ def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[str, Data
         self.logger.debug(f"[{total_time:.4f}] Processing batch of size %s into %s profiles", len(batch), len(results))
 
 
+class SegmentedProfileViews(beam.DoFn):
+    def __init__(self, args: TemplateArgs):
+        self.date_column = args.date_column
+        self.freq = args.date_grouping_frequency
+        self.logging_level = args.logging_level
+        self.logger = logging.getLogger("ProfileViews")
+        self.segment_columns = args.segment_columns
+
+    def setup(self) -> None:
+        self.logger.setLevel(logging.getLevelName(self.logging_level))
+
+    @beam.DoFn.yields_elements
+    def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[SegmentDefinition, DatasetProfileView]]:
+        # If no segment column is informed, no point in running the processing
+        assert self.segment_columns is not None
+
+        start_time = time.perf_counter()
+        tmp_date_col = "_whylogs_datetime"
+        df = pd.DataFrame(batch)
+        df[tmp_date_col] = pd.to_datetime(df[self.date_column])
+        grouped = df.set_index(tmp_date_col).groupby(pd.Grouper(freq=self.freq))
+
+        self.logger.info(f"Using {self.segment_columns} for segmentation")
+
+        seg_columns = self.segment_columns.split(",")
+
+        for segment_column in seg_columns:
+            # trim whitespaces on every string
+            segment_column = segment_column.strip()
+
+            column_segments = segment_on_column(segment_column)
```

Comment:
Is this the right way of doing that? I thought you could just give a list of segments to whylogs and it would create the segments it needs to all in one go, and then you could get them all out at once. With this approach we end up processing the entire batch once for each segment defined.

Reply:
As a side note, we'll have to put a big warning in the docs about how this can blow up, even if we can figure out how to limit it to a single pass. Does whylogs throw or something if this is a continuous column that would explode the number of segment partitions?